src/api.ts

```diff
@@ -822,6 +960,47 @@ function createApi({
         )
       }
 
+      let spriteId: string | undefined
+
+      if (style.sprite) {
+        let upstreamSprites:
+          | Awaited<ReturnType<typeof fetchUpstreamSprites>>['sprites']
+          | undefined
+        try {
+          const result = await fetchUpstreamSprites({
+            accessToken,
+            upstreamSpriteUrl: style.sprite,
+          })
+
+          if (
+            [...result.sprites.values()].every((info) => info instanceof Error)
+          ) {
+            throw new FailedUpstreamFetchError(style.sprite)
+          }
+
+          spriteId = result.id
+          upstreamSprites = result.sprites
+        } catch (err) {
+          // This will either be an error about the access token or the inability to fetch any sprites upstream
+          throw err
+        }
+
+        // TODO: Wrap in transaction?
+        for (const [pixelDensity, spriteInfo] of upstreamSprites.entries()) {
+          // TODO: Should we report the error here?
+          if (spriteInfo instanceof Error) continue
+
+          await api.createSprite({
+            id: spriteId,
+            data: spriteInfo.data,
+            etag: spriteInfo.etag || null,
+            layout: JSON.stringify(spriteInfo.layout),
+            pixelDensity,
+            upstreamUrl: style.sprite,
+          })
+        }
+      }
+
```

Comment:
Coming back to this, I understand why you have this, in order to have a spriteId for inserting the style. It makes me wonder whether we should instead hash the sprite URL as the spriteId? or even just use the actual URL as the id, so that we could do something similar to what we do with tiles: receive a request for a spriteId, if it doesn't exist locally, then decode the upstream URL from the spriteId, and request from upstream and populate the table. With the Mapbox API the URL changes every time the sprite changes. For other map APIs I assume we could rely on the etag as an indicator if we need to re-fetch.

Reply:
not sure using just the plain sprite url for the sprite id would work because it's directly inserted into the `sprite` field for when a style is requested e.g. if the sprite url is something like `mapbox://sprites/user/spritename` then our generated field value would be `base/styles/some-style-id/sprites/mapbox://sprites/user/spritename`, which seems problematic as a url?

Hashing the sprite url seems like an option though

Reply:
> I'm not sure (actually not sure) whether this is the best place for this? I'm thinking of when we add the option to import a styled map from a file/folder, and in that case we would want to create the style and the sprites separately (using the files that we are importing).

Yeah that makes sense! I think that's probably a better way of going about it i.e. in the route function we can follow approach:

1. attempt to fetch and create sprites
2. create style, populating sprite id derived from step 1

or 

1. create style
2. attempt to fetch and create sprites
3. update the created style to have the sprite id derived from 2

Reply:
implemented what turns out to be a combination of both approaches in 9f7fc47f45518d4e8d9851a105576cf11fe5e3af:


1. Attempt to fetch sprites (if necessary)
2. Create style (uses sprite id generated from hashing sprite url)
3. Create sprites (if necessary)